### PR TITLE
feat(c): support creating nested values in C API

### PR DESCRIPTION
**Pull request by Mause**
Pretty open to feedback here - while this feels like the simplest interface, it probably isn't the most efficient (the alternative is having a "builder" with the initial allocation call, and `duckdb_set_value(duckdb_value parent, int index, duckdb_value child)` or something

In part fixes https://github.com/duckdb/duckdb/discussions/9365

### CODE REVIEW

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! LGTM - comments below:

File: src/main/capi/duckdb_value-c.cpp [hunk 1]
```diff
@@ -39,3 +53,45 @@ int64_t duckdb_get_int64(duckdb_value value) {
...
+	}
+	auto ltype = UnwrapType(type);
+	duckdb::vector<Value> unwrapped_values = duckdb::vector<Value>();
+
+	for (idx_t i = 0; i < value_count; i++) {
+		auto val = UnwrapValue(&values[i]);
+		if (!val) {
+			return nullptr;
+		}
+		unwrapped_values.push_back(*val);
```
[the remark this answers is not in the document]
Comment (Mause) [reply src/main/capi/duckdb_value-c.cpp]: For now, it'll try to cast it, or return an error (ie, null)

File: src/main/capi/duckdb_value-c.cpp [hunk 2]
```diff
@@ -39,3 +53,45 @@ int64_t duckdb_get_int64(duckdb_value value) {
...
+	auto &struct_type = duckdb::StructType::GetChildTypes(ltype);
+	auto count = struct_type.size();
+
+	duckdb::child_list_t<Value> unwrapped_values;
+	for (idx_t i = 0; i < count; i++) {
+		auto val = UnwrapValue(&values[i]);
+		if (!val) {
+			return nullptr;
+		}
+		unwrapped_values.emplace_back(struct_type.get(i).first, *val);
```
[the remark this answers is not in the document]
Comment (Mause) [reply src/main/capi/duckdb_value-c.cpp]: For now, it'll try to cast it, or return an error (ie, null)

**Comment by Mytherin** [comment]
Thanks!

File: src/common/types/value.cpp [hunk 3]
```diff
@@ -644,18 +644,25 @@ Value Value::TIMESTAMP(int32_t year, int32_t month, int32_t day, int32_t hour, i
 	return val;
 }
 
-Value Value::STRUCT(child_list_t<Value> values) {
+Value Value::STRUCT(const LogicalType &type, vector<Value> struct_values) {
 	Value result;
+	auto child_types = StructType::GetChildTypes(type);
```
Comment (Tishj) [inline src/common/types/value.cpp:649]: A little late to review this, my bad, but still might be worthwhile to mention
Shouldn't we assert that `struct_values.size()` == `child_types.size()`

File: src/main/capi/duckdb_value-c.cpp [hunk 4]
```diff
@@ -1,9 +1,22 @@
 #include "duckdb/main/capi/capi_internal.hpp"
+#include "duckdb/common/types/value.hpp"
+#include "duckdb/common/types.hpp"
 
+static duckdb_value WrapValue(duckdb::Value *list_value) {
```
Comment (Tishj) [inline src/main/capi/duckdb_value-c.cpp:5]: If this takes a `unique_ptr<duckdb::Value> &&value` we can better safeguard that `duckdb_value` doesn't contain stack memory.

File: src/main/capi/duckdb_value-c.cpp [hunk 5]
```diff
@@ -39,3 +52,54 @@ int64_t duckdb_get_int64(duckdb_value value) {
...
+		unwrapped_values.emplace_back(UnwrapValue(value));
+	}
+	duckdb::Value *struct_value = new duckdb::Value;
+	try {
+		*struct_value = duckdb::Value::STRUCT(ltype, unwrapped_values);
+	} catch (...) {
+		delete struct_value;
+		return nullptr;
+	}
+	return WrapValue(struct_value);
```
Comment (Tishj) [inline src/main/capi/duckdb_value-c.cpp:81]: When this takes a unique_ptr, we also don't need to explicitly use `new` and `delete` here

**Pull request merged**